README.md

# s3labeler  

s3 - Amazon Simple Storage Service    
label - attached to an object and giving information about it    

## S3 Object Labeling Tool   

[![Package Version](https://img.shields.io/pypi/v/s3labeler.svg)](https://pypi.python.org/pypi/s3labeler/)
[![Python 3.6](https://img.shields.io/badge/python-3.6-blue.svg)](https://www.python.org/downloads/release/python-360/)


## install
```
pip install s3labeler
```

## cli  
```
s3labeler --help
```

```
Usage: ./s3labeler.py option

    options:

        list-buckets|buckets

        list|ls    <s3bucket>/<s3object>
        label|set  <s3bucket>/<s3object> '{"label":"value"}'
        delete|del <s3bucket>/<s3object> label

        get    <s3bucket>/<s3object>
        save   <s3bucket>/<s3object> destination
        upload source <s3bucket>/<s3object>

        rekognition <s3bucket>/<s3object>
        rekognition <s3bucket>/<s3object> detect-labels
        rekognition <s3bucket>/<s3object> words
        rekognition <s3bucket>/<s3object> s3tag

        rekognition <s3bucket>/<s3object> confidence
        rekognition <s3bucket>/<s3object> s3tag confidence top 3

        object      <s3bucket>/<s3object>
        b2sum       <s3bucket>/<s3object>
        identify|id <s3bucket>/<s3object>

        server 8880

        --help
        --version

```

## run as a service
```
s3labeler server
```

## s3 rest api

#### list all buckets HTTP GET
```
curl http://127.0.0.1:8880/s3/ 
```

#### list files in a bucket (1000 record limit) HTTP GET
```
curl http://127.0.0.1:8880/s3/<s3bucket>/
```

#### list files in a bucket subdirectory (1000 record limit) HTTP GET
```
curl http://127.0.0.1:8880/s3/<s3bucket>/rekognition/
```

#### list s3objects s3tags HTTP GET
```
curl http://127.0.0.1:8880/s3/<s3bucket>/<s3object>
```

#### view s3object content, specify content-type header HTTP GET
```
curl http://127.0.0.1:8880/s3/<s3bucket>/<s3object>?image=jpeg
```
image/gif image/jpeg image/png image/tiff image/vnd.microsoft.icon image/x-icon image/vnd.djvu image/svg+xml  


#### list s3object s3tags HTTP GET (same as default get)
```
curl "http://127.0.0.1:8880/s3/<s3bucket>/<s3object>?tags=s3"
```

#### list file rekognition json HTTP GET
```
curl "http://127.0.0.1:8880/s3/<s3bucket>/<s3object>?tags=rekognition"
```

#### set s3object tag HTTP GET
```   
curl -X GET "http://127.0.0.1:8880/s3/<s3bucket>/<s3object>?label=tag_name&value=something"
```

#### delete s3object tag HTTP GET
```   
curl -X GET "http://127.0.0.1:8880/s3/<s3bucket>/<s3object>?delete=tag_name"
```

#### set or update s3object tag HTTP PATCH (single key/value)
```   
curl -X PATCH \
     -H "Content-Type: application/json" \
     -d '{"labeler":"karl"}' \
     http://127.0.0.1:8880/s3/<s3bucket>/<s3object>    
```   

#### delete s3object tag HTTP DELETE
```   
curl -X DELETE "http://127.0.0.1:8880/s3/<s3bucket>/<s3object>?tag=tag_name"
```   


#### set s3object tag set HTTP PUT (Warning: this method overwrites the entire tagset)
```   
curl -X PUT \
     -H "Content-Type: application/json" \
     -d '{"labeler":"karl","image_url":"https://<s3bucket>.s3.us-west-2.amazonaws.com/<s3object>"}' \
     http://127.0.0.1:8880/s3/<s3bucket>/<s3object>    
```   

### HTTP POST
two available endpoints
```
/s3/<s3bucket>/<s3object>  
```
and/or
```
/api
```

http post json or form data to s3object endpoint

#### set or update s3object tag HTTP POST (single key/value) s3object endpoint
```   
curl -X POST \
     -H "Content-Type: application/json" \
     -d '{"labeler":"karl"}' \
     http://127.0.0.1:8880/s3/<s3bucket>/<s3object>
```
```
    <form action="http://127.0.0.1:8880/s3/<s3bucket>/<s3object>" method="POST">
      <input type="text" name="label" value="label">
      <input type="text" name="value" value="value">
      <input type="submit">
    </form>
```

http post json or form data to api endpoint

#### set or update s3object tag HTTP POST /api endpoint
```   
curl -X POST \
     -H "Content-Type: application/json" \
     -d '{"s3bucket":"<s3bucket>","s3object":"<s3object>","labeler":"karl"}' \
     http://127.0.0.1:8880/api
```
```
    <form action="http://127.0.0.1:8880/api" method="POST">
        <input type="text" name="s3bucket" value="s3bucket">
        <input type="text" name="s3object" value="s3object">
        <input type="text" name="label" value="label">
        <input type="text" name="value" value="value">
        <input type="submit">
    </form>
```

---

#### get rekognition json HTTP GET (same as ?tags=rekognition)
```
curl "http://127.0.0.1:8880/s3/<s3bucket>/<s3object>?rekognition=json"
```

#### get rekognition words HTTP GET
```
curl "http://127.0.0.1:8880/s3/<s3bucket>/<s3object>?rekognition=words"
```
 
#### set rekognition words to s3object tag
```
curl "http://127.0.0.1:8880/s3/<s3bucket>/<s3object>?rekognition=words&save=s3tag"
```

#### get rekognition confidence HTTP GET
```
curl "http://127.0.0.1:8880/s3/<s3bucket>/<s3object>?rekognition=confidence"
```

#### get rekognition confidence top 3 HTTP GET
```
curl "http://127.0.0.1:8880/s3/<s3bucket>/<s3object>?rekognition=confidence&top=3"
```
#### get rekognition confidence top 90 percent HTTP GET
```
curl "http://127.0.0.1:8880/s3/<s3bucket>/<s3object>?rekognition=confidence&top=90&percent=true"
```

#### set rekognition confidence to s3object tag (all confidence words)
```
curl "http://127.0.0.1:8880/s3/<s3bucket>/<s3object>?rekognition=confidence&save=s3tag"
```
#### set rekognition confidence to s3object tag (top 3 confidence)
```
curl "http://127.0.0.1:8880/s3/<s3bucket>/<s3object>?rekognition=confidence&top=3&save=s3tag"
```
#### set rekognition confidence to s3object tag (top 90 precent)
```
curl "http://127.0.0.1:8880/s3/<s3bucket>/<s3object>?rekognition=confidence&top=90&percent=true&save=s3tag"
```

### run image through aws rekognition

#### run image through aws rekognition detect-labels HTTP GET  
```
curl "http://127.0.0.1:8880/s3/<s3bucket>/<s3object>?rekognition=detect-labels"
```

#### run image through aws rekognition detect-labels and save/overwrite with new json HTTP GET
```
curl "http://127.0.0.1:8880/s3/<s3bucket>/<s3object>?rekognition=detect-labels&save=true"
```

---

## run from source
```
python3 src/s3labeler/s3labeler.py
```

## run as python module
```
cd src/ && python3 -m s3labeler
```

---

image label manager   
run an image through rekognition only once (to prevent recurring costs).  update s3objects tags.   

---

tech notes   
 - aws rekognition detect-labels leverages amazons proprietary ML model for image recognition   
 - image file bytes are not transfered or processed through this interface for rekognition     

https://aws.amazon.com/s3/   
https://aws.amazon.com/rekognition/    







CHANGES
1.0.5 - https://flask.palletsprojects.com/en/2.0.x/patterns/appfactories/
1.0.4 - confidence
1.0.3 - flask-cors
1.0.2 - http.get ?delete=label, ?label=name&value=something
1.0.1 - http.post
1.0.0 - initial release 


src/s3_labeler/t/t2.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import boto3

#s3 = boto3.resource('s3')

#conn = boto3.client('s3')

#my_bucket = s3.Bucket('ninfo-property-images/rekognition/')
#for my_bucket_object in my_bucket.objects.all():
#    print(my_bucket_object)

#for key in conn.list_objects(Bucket='ninfo-property-images')['rekognition']:
#    print(key['Key'])

#    from boto3  import client
#    bucket_name = "my_bucket"
#    prefix      = "my_key/sub_key/lots_o_files"
#
#    s3_conn   = client('s3')  # type: BaseClient  ## again assumes boto.cfg setup, assume AWS S3
#    s3_result =  s3_conn.list_objects_v2(Bucket=bucket_name, Prefix=prefix, Delimiter = "/")
#
#    if 'Contents' not in s3_result:
#        #print(s3_result)
#        return []
#

bucket      = "ninfo-property-images"
prefix      = "rekognition/"

s3_conn   = boto3.client('s3')
s3_result =  s3_conn.list_objects_v2(Bucket=bucket, Prefix=prefix, Delimiter = "/")

for key in s3_result['Contents']:
    print(key['Key'])















src/s3_labeler/t/t3.py

def stat(bucket, key):
    try:
        client = boto3.client("s3")
        return client.head_object(Bucket=bucket, Key=key)

    except botocore.exceptions.ClientError as err:
        status = err.response["ResponseMetadata"]["HTTPStatusCode"]
        errcode = err.response["Error"]["Code"]
        if status == 404:
            logging.warning("Missing object, %s", errcode)
        elif status == 403:
            logging.error("Access denied, %s", errcode)
        else:
            logging.exception("Error in request, %s", errcode)
    return {}





src/s3labeler/t/bototag.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import boto3
import sys


s3bucket = sys.argv[1]
s3object = sys.argv[2]
tagkey   = sys.argv[3]
tagval   = sys.argv[4]

print(s3bucket)
print(s3object)

s3_client = boto3.client('s3',
    #region_name='region-name',
    #aws_access_key_id='aws-access-key-id',
    #aws_secret_access_key='aws-secret-access-key',
)

get_tags_response = s3_client.get_object_tagging(
    Bucket=s3bucket,
    Key=s3object,
)

put_tags_response = s3_client.put_object_tagging(
    Bucket=s3bucket,
    Key=s3object,
    Tagging={
        'TagSet': [
            {
                'Key': 'labeler',
                'Value': 'karl.rink'
            },
        ]
    }
)






#s3 = boto3.resource('s3')
#https://boto.readthedocs.io/en/latest/ref/s3.html#boto.s3.bucket.Bucket.set_tags
#https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/s3.html#S3.Client.put_object_tagging







src/image_labeler/s3imglbr.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

__version__ = '0.0.0.a5'

import sys

if sys.version_info < (3, 8, 1):
    raise RuntimeError('Requires Python version 3.8.1 or higher. This version: ' + str(sys.version_info))

from flask import Flask
from flask import request
from flask import jsonify
from werkzeug.exceptions import HTTPException

app = Flask(__name__)
app.config['JSONIFY_PRETTYPRINT_REGULAR'] = True    #default False
app.config['JSON_SORT_KEYS'] = True                 #default True
app.config['JSONIFY_MIMETYPE'] = 'application/json' #default 'application/json'

import boto3
import botocore
#from botocore.exceptions import ClientError

import json

#GET    /                             # Show version
@app.route("/", methods=['GET'])
def root():
    return jsonify(status=200, message="OK", version=__version__), 200, {'Content-Type':'application/json;charset=utf-8'}


#GET    /s3                           # Show OK
@app.route("/s3", methods=['GET'])
def get_s3():
    return jsonify(status=200, message="OK", path="/s3"), 200, {'Content-Type':'application/json;charset=utf-8'}


#GET    /s3/                          # List all buckets (limt 1000?)
@app.route("/s3/", methods=['GET'])   # fun, this method has a "Bucket List"
def get_s3buckets(region=None):

    s3 = boto3.resource('s3', region_name=region)

    bucket_list = [b.name for b in s3.buckets.all()]

    #for bucket in bucket_list:
    #    print(bucket)

    #return jsonify(status=200, message="OK", success=True), 200
    return jsonify(bucket_list), 200, {'Content-Type':'application/json;charset=utf-8'}



#GET    /s3/<s3bucket>/<s3object>     # List object
#GET    /s3/<s3bucket>/<s3object>?q=  # rekognition=json|words|detect-labels
#GET    /s3/<s3bucket>/<s3object>?q=  # tags=s3|rekognition

#GET    /s3/<s3bucket>/<s3object>?q=
@app.route("/s3/<s3bucket>/<s3object>", methods=['GET'])
def get_s3bucketobject(s3bucket=None,s3object=None):

    assert s3bucket == request.view_args['s3bucket']
    assert s3object == request.view_args['s3object']

    rekognition = request.args.get("rekognition", None)
    tags        = request.args.get("tags", None)

    s3_client = boto3.client('s3')

    try:
        s3_result = s3_client.list_objects_v2(Bucket=s3bucket, Prefix=s3object, Delimiter = "/")
    except botocore.exceptions.EndpointConnectionError as e:
        return jsonify(status=599, message="EndpointConnectionError", error=str(e)), 599, {'Content-Type':'application/json;charset=utf-8'}

    try:
        for key in s3_result['Contents']:
            #print(key['Key'])
            _k = key['Key']
    except KeyError as e:
        _k = s3_result['Prefix']
        return jsonify(status=404, message="Not Found", s3object=False, name=_k), 404, {'Content-Type':'application/json;charset=utf-8'}

    if _k != s3object:
        return jsonify(status=569, message="Objects Do Not Match", object1=str(_k), ojbect2=str(s3object)), 569, {'Content-Type':'application/json;charset=utf-8'}

    #s3object#

    if tags:

        if tags == 's3':
            print('s3tags')
            #aws s3api get-object-tagging --bucket ninfo-property-images --key 2eece964b6f902124052810e5a92d6f9ca715c1b.jpg

            get_s3tags = get_s3object_tags(s3bucket, s3object)
            #s3tags = get_s3object_tags(s3bucket, s3object)
            #print(s3tags)

            #for key in s3tags['TagSet']:
            #    #print(key)
            #    print(key['Key'], key['Value'])

            #print(str(s3tags['TagSet']))

            s3Tags = {}
            for key in get_s3tags['TagSet']:
                __k = key['Key']
                __v = key['Value']
                s3Tags[__k]=__v

            return jsonify(s3Tags), 200, {'Content-Type':'application/json;charset=utf-8'}


        if tags == 'rekognition':

            rekognition_json_file = 'rekognition/' + s3object + '.json'

            rekognition_json_content = get_rekognition_json(s3bucket, rekognition_json_file)

            if rekognition_json_content:
                return rekognition_json_content, 200, {'Content-Type':'application/json;charset=utf-8'}
            else:
                return jsonify(status=404, message="Not Found", s3object=False, rekognition_json_location=rekognition_json_file), 404, {'Content-Type':'application/json;charset=utf-8'}


    if rekognition:

        if rekognition == 'json':
            rekognition_json_file = 'rekognition/' + s3object + '.json'
            rekognition_json_content = get_rekognition_json(s3bucket, rekognition_json_file)
            if rekognition_json_content:
                return rekognition_json_content, 200, {'Content-Type':'application/json;charset=utf-8'}
            else:
                return jsonify(status=404, message="Not Found", s3object=False, rekognition_json_location=rekognition_json_file), 404, {'Content-Type':'application/json;charset=utf-8'}

        if rekognition == 'words':

            rekognition_json_file = 'rekognition/' + s3object + '.json'
            rekognition_json_content = get_rekognition_json(s3bucket, rekognition_json_file)

            #print(str(type(rekognition_json_content)))
            #print(str(rekognition_json_content))

            #_wL = extract_rekognition_words(rekognition_json_content)
            #print(str(_wL))

            if rekognition_json_content:

                #wordList = [ 'Tree', 'Garage', 'Donuts' ]

                wordList = extract_rekognition_words(rekognition_json_content)
                print(str(wordList))

                return jsonify(wordList), 200, {'Content-Type':'application/json;charset=utf-8'}
            else:
                return jsonify(status=404, message="Not Found", s3object=False, rekognition_json_location=rekognition_json_file), 404, {'Content-Type':'application/json;charset=utf-8'}


            #wordList = [ 'Tree', 'Garage', 'Donuts' ]
            #if len(wordList) > 0:
            #    return jsonify(wordList), 200, {'Content-Type':'application/json;charset=utf-8'}
            #else:
            #    return jsonify(status=404, message="Not Found", s3object=False, rekognition_json_location=rekognition_json_file), 404, {'Content-Type':'application/json;charset=utf-8'}


    return jsonify(status=200, message="OK", s3object=True, name=_k), 200, {'Content-Type':'application/json;charset=utf-8'}



##GET    /s3/<s3bucket>/<s3object>     # List object
#@app.route("/s3/<s3bucket>/<s3object>", methods=['GET'])
#def get_s3bucketobject(s3bucket=None,s3object=None):
#    assert s3bucket == request.view_args['s3bucket']
#    assert s3object == request.view_args['s3object']
#
#    s3_client = boto3.client('s3')
#
#    try:
#        s3_result = s3_client.list_objects_v2(Bucket=s3bucket, Prefix=s3object, Delimiter = "/")
#    except botocore.exceptions.EndpointConnectionError as e:
#        return jsonify(status=599, message="EndpointConnectionError", error=str(e)), 599
#
#    try:
#        for key in s3_result['Contents']:
#            #print(key['Key'])
#            _k = key['Key']
#    except KeyError as e:
#        _k = s3_result['Prefix']
#        return jsonify(status=404, message="Not Found", existing=False, key=_k), 404
#
#    if _k != s3object:
#        return jsonify(status=569, message="Objects Do Not Match", object1=str(_k), ojbect2=str(s3object)), 569
#
#    return jsonify(status=200, message="OK", existing=True, key=_k), 200


#PUT      /s3/<s3bucket>/<s3object>         # set s3object tag set keys and values   
@app.route("/s3/<s3bucket>/<s3object>", methods=['PUT'])
def set_s3bucketobject(s3bucket=None,s3object=None):

    assert s3bucket == request.view_args['s3bucket']
    assert s3object == request.view_args['s3object']

    if not request.headers['Content-Type'] == 'application/json':
        return jsonify(status=412, errorType="Precondition Failed"), 412

    post = request.get_json()

    #print(str(type(post))) #<class 'dict'>
    #print(str(post))       #{'labler': 'karl...
    
    #jsondata = json.loads('{}')

    #settagset = set_s3object_tags(s3bucket, s3object, jsondata)
    settagset = set_s3object_tags(s3bucket, s3object, post)

    print(settagset)

    return jsonify(status=200, message="OK", name=s3object, method="PUT"), 200, {'Content-Type':'application/json;charset=utf-8'}


#PATCH    /s3/<s3bucket>/<s3object>         # set s3object tag    
@app.route("/s3/<s3bucket>/<s3object>", methods=['PATCH'])
def set_s3bucketobjectpatch(s3bucket=None,s3object=None):

    assert s3bucket == request.view_args['s3bucket']
    assert s3object == request.view_args['s3object']

    if not request.headers['Content-Type'] == 'application/json':
        return jsonify(status=412, errorType="Precondition Failed"), 412

    post = request.get_json()

    if len(post) > 1:
        return jsonify(status=405, errorType="Method Not Allowed", errorMessage="Single Key-Value Only", update=False), 405

    print(post)
    for k,v in post.items():
        tag=k
        value=v

    update = update_s3object_tag(s3bucket, s3object, tag, value)
   
    print(update)

    if update is True:
        return jsonify(status=200, message="OK", name=s3object, method="PATCH"), 200, {'Content-Type':'application/json;charset=utf-8'}

    #return jsonify(status=200, message="OK", name=s3object, method="PUT"), 200, {'Content-Type':'application/json;charset=utf-8'}
    return jsonify(status=465, message="Failed Patch", name=s3object, tag=tag, method="PATCH", update=False), 465, {'Content-Type':'application/json;charset=utf-8'}




#DELETE      /s3/<s3bucket>/<s3object>?tag=name      # delete s3object tag 
@app.route("/s3/<s3bucket>/<s3object>", methods=['DELETE'])
def delete_s3bucketobjecttag(s3bucket=None,s3object=None):

    assert s3bucket == request.view_args['s3bucket']
    assert s3object == request.view_args['s3object']

    tag = request.args.get("tag", None)

    if tag:
        #return jsonify(status=466, message="No Such Tag", name=s3object, tag=tag, method="DELETE", delete=False), 466, {'Content-Type':'application/json;charset=utf-8'}
        #return jsonify(status=466, message="No Such Tag", name=s3object, tag=tag, method="DELETE", delete=False), 466, {'Content-Type':'application/json;charset=utf-8'}

        delete_tag = delete_s3object_tag(s3bucket, s3object, tag)

        if delete_tag is True:

            return jsonify(status=211, message="Deleted", name=s3object, tag=tag, method="DELETE", delete=True), 211, {'Content-Type':'application/json;charset=utf-8'}


    #return jsonify(status=211, message="Deleted", name=s3object, tag=tag, method="DELETE", delete=True), 211, {'Content-Type':'application/json;charset=utf-8'}
    return jsonify(status=466, message="Failed Delete", name=s3object, tag=tag, method="DELETE", delete=False), 466, {'Content-Type':'application/json;charset=utf-8'}




#GET    /s3/<s3bucket>/<s3subdir>/<s3object> # List object
@app.route("/s3/<s3bucket>/<s3subdir>/<s3object>", methods=['GET'])
def get_s3bucketsubdirobject(s3bucket=None,s3subdir=None,s3object=None):
    assert s3bucket == request.view_args['s3bucket']
    assert s3subdir == request.view_args['s3subdir']
    assert s3object == request.view_args['s3object']

    #print(s3bucket)
    #print(s3subdir)
    #print(s3object)

    s3_client = boto3.client('s3')

    prefix = s3subdir + '/' + s3object

    s3_result = s3_client.list_objects_v2(Bucket=s3bucket, Prefix=prefix, Delimiter = "/")

#    for key in s3_result['Contents']:
#        print(key['Key'])

    _exist=False

    try:
        for key in s3_result['Contents']:
            #print(key['Key'])
            _k = key['Key']
            _exist=True
    except KeyError as e:
        _k = s3_result['Prefix']
        return jsonify(status=404, message="Not Found", s3object=_exist, name=_k), 404, {'Content-Type':'application/json;charset=utf-8'}

    return jsonify(status=200, message="OK", s3object=_exist, key=_k), 200, {'Content-Type':'application/json;charset=utf-8'}


#GET    /s3/<s3bucket>/<s3subdir>/                         # List bucket directory files (1000 limit)
@app.route("/s3/<s3bucket>/<s3subdir>/", methods=['GET'])
def get_s3bucketsubdir(s3bucket=None,s3subdir=None):       
    assert s3bucket == request.view_args['s3bucket']
    assert s3subdir == request.view_args['s3subdir']

    #print(s3bucket)
    #print(s3subdir)

    #bucket      = "bucket-name"
    #prefix      = "folder/"

    prefix = s3subdir + '/'

    s3_client = boto3.client('s3')
    s3_result =  s3_client.list_objects_v2(Bucket=s3bucket, Prefix=prefix, Delimiter = "/")

    s3objects=[]
    try:
        for key in s3_result['Contents']:
            #print(key['Key'])
            s3objects.append(key['Key'])
    except KeyError as e:
        return jsonify(status=404, message="Not Found", prefix=prefix), 404, {'Content-Type':'application/json;charset=utf-8'}

    print(len(s3objects))
    if len(s3objects) < 0:
        return jsonify(status=404, message="Not Found", s3object=False, prefix=prefix), 404, {'Content-Type':'application/json;charset=utf-8'}

    #return jsonify(status=200, message="OK", existing=_exist, key=None), 200
    #return jsonify(status=200, message="OK", prefix=prefix), 200
    return jsonify(s3objects), 200, {'Content-Type':'application/json;charset=utf-8'}
     

#GET    /s3/<s3bucket>/                         # List bucket directory files (1000 limit)
@app.route("/s3/<s3bucket>/", methods=['GET'])
def get_s3bucketdir(s3bucket=None):             
    assert s3bucket == request.view_args['s3bucket']

    #print(s3bucket)

    s3_client = boto3.client('s3')

    try:
        s3_result =  s3_client.list_objects_v2(Bucket=s3bucket, Delimiter = "/")
    except s3_client.exceptions.NoSuchBucket as e:
        return jsonify(status=404, message="No Such Bucket", s3bucket=s3bucket), 404, {'Content-Type':'application/json;charset=utf-8'}

    s3objects=[]
    try:
        for key in s3_result['Contents']:
            s3objects.append(key['Key'])
    except KeyError as e:
        return jsonify(status=404, message="Not Found", s3bucket=s3bucket), 404, {'Content-Type':'application/json;charset=utf-8'}

    #print(len(s3objects))
    if len(s3objects) < 0:
        return jsonify(status=404, message="Not Found", s3objects=0, s3bucket=s3bucket), 404, {'Content-Type':'application/json;charset=utf-8'}

    return jsonify(s3objects), 200, {'Content-Type':'application/json;charset=utf-8'}


@app.errorhandler(Exception)
def handle_exception(e):

    if isinstance(e, HTTPException):
        return jsonify(status=e.code, errorType="HTTPException", errorMessage=str(e)), e.code, {'Content-Type':'application/json;charset=utf-8'}

    return jsonify(status=599, errorType="Exception", errorMessage=str(e)), 599, {'Content-Type':'application/json;charset=utf-8'}


@app.errorhandler(404)
def not_found(error=None):
    message = { 'status': 404, 'errorType': 'Not Found: ' + request.url }
    return jsonify(message), 404, {'Content-Type':'application/json;charset=utf-8'}



#@app.errorhandler(Exception)
#def handle_exception(e):
#
#    if isinstance(e, HTTPException):
#        return jsonify(status=e.code, errorType="HTTP Exception", errorMessage=str(e)), e.code
#
#    if type(e).__name__ == 'OperationalError':
#        return jsonify(status=512, errorType="OperationalError", errorMessage=str(e)), 512
#
#    if type(e).__name__ == 'InterfaceError':
#        return jsonify(status=512, errorType="InterfaceError", errorMessage=str(e)), 512
#
#    if type(e).__name__ == 'ProgrammingError':
#        return jsonify(status=512, errorType="ProgrammingError", errorMessage=str(e)), 512
#
#    if type(e).__name__ == 'AttributeError':
#        return jsonify(status=512, errorType="AttributeError", errorMessage=str(e)), 512
#
#    res = {'status': 500, 'errorType': 'Internal Server Error'}
#    res['errorMessage'] = str(e)
#    return jsonify(res), 500
#


###############################################################################################################################################

def get_rekognition_json(s3bucket, s3object):

    #rekognition_json_file = 'rekognition/' + s3object + '.json'

    #s3 = boto3.resource('s3')
    #getobject = s3.Object('bucket_name','key')
    #getobject = s3.Object(s3bucket, s3object)
    #print(str(type(getobject))) #<class 'boto3.resources.factory.s3.Object'>
    #print(str(getobject))       #s3.Object(bucket_name='ninfo-property-images', key='2eece964b6f902124052810e5a92d6f9ca715c1b.jpg')
    #return getobject
    #obj = s3.Object(s3bucket, s3object)
    #body = obj.get()['Body'].read()

    #return get_s3object_body(s3bucket, rekognition_json_file)
    return get_s3object_body(s3bucket, s3object)


def get_s3object_body(s3bucket, s3object): #gets file contents data
    s3 = boto3.resource('s3')
    obj = s3.Object(s3bucket, s3object)
    try:
        body = obj.get()['Body'].read()
    except botocore.exceptions.ClientError as ex:
        print('ClientError ' + str(ex))
        #if ex.response['Error']['Code'] == 'NoSuchKey':
        #    print('NoSuchKey')
        #    return None
        return None

    #return body                #<class 'bytes'>
    return body.decode('utf-8') #<class 'str'>

def get_s3object_tags(s3bucket, s3object):
    s3_client = boto3.client('s3')
    s3_result = s3_client.get_object_tagging(Bucket=s3bucket, Key=s3object)
    return s3_result

def set_s3object_tags(s3bucket, s3object, post):
    """ multi key/value """

    s3_client = boto3.client('s3')

    #print(str(post))

    #s3_result = s3_client.put_object_tagging(
    #        Bucket=s3bucket,
    #        Key=s3object,
    #        Tagging={'TagSet':[]}
    #        )

    # {},{} #nested python dict

    #_Ts = { 'TagSet': [ {'Key': 'tag-key', 'Value': 'tag-value'} ] }
    #_Ts = { 'TagSet': [ {'Key': 'tag-key', 'Value': 'tag-value'}, {'Key': 'tag-key2', 'Value': 'tag-value2'} ] }

    KeyValList = []

    for k,v in post.items():
        #print(' jpost ' + k,v)
        kvs={}
        kvs['Key']   = k
        kvs['Value'] = v

        KeyValList.append(kvs)

    _Ts = { 'TagSet': KeyValList }
    print(_Ts)

    s3_result = s3_client.put_object_tagging(
            Bucket=s3bucket,
            Key=s3object,
            Tagging=_Ts
            )

    return None

def update_s3object_tag(s3bucket, s3object, tag, value):
    """ single key/value """

    s3_client = boto3.client('s3')

    get_tags_response = s3_client.get_object_tagging(
        Bucket=s3bucket,
        Key=s3object,
    )

    # get existing tags
    s3Tags = {}
    for key in get_tags_response['TagSet']:
        __k = key['Key']
        __v = key['Value']
        s3Tags[__k]=__v

    #overwrite existing key/value w/ new 
    s3Tags[tag]=value

    #write out new dict
    KeyValList = []
    for k,v in s3Tags.items():
        kvs={}
        kvs['Key']   = k
        kvs['Value'] = v
        KeyValList.append(kvs)

    TagSet = { 'TagSet': KeyValList }

    put_tags_response = s3_client.put_object_tagging(
        Bucket=s3bucket,
        Key=s3object,
        Tagging=TagSet
    )

    #print(str(put_tags_response))

    #for key in put_tags_response['ResponseMetadata']:

    status_code = put_tags_response['ResponseMetadata']['HTTPStatusCode']

    if int(status_code) == 200:
        return True
    else:
        return False




#put_tags_response = s3_client.put_object_tagging(
#    Bucket='your-bucket-name',
#    Key='folder-if-any/file-name.extension',
#    Tagging={
#        'TagSet': [
#            {
#                'Key': 'tag-key',
#                'Value': 'tag-value'
#            },
#        ]
#    }
#)

def delete_s3object_tag(s3bucket, s3object, tag):
    
    s3_client = boto3.client('s3',
    #region_name='region-name',
    #aws_access_key_id='aws-access-key-id',
    #aws_secret_access_key='aws-secret-access-key',
    )

    get_tags_response = s3_client.get_object_tagging(
        Bucket=s3bucket,
        Key=s3object,
    )

    # get existing tags
    s3Tags = {}
    for key in get_tags_response['TagSet']:
        __k = key['Key']
        __v = key['Value']
        s3Tags[__k]=__v

    #overwrite existing key/value w/ new 
    #s3Tags[tagkey]=tagval

    # delete the key regardless
    #s3Tags.pop(tag, None)

    if tag not in s3Tags:
        return False

    # delete the key regardless
    delete = s3Tags.pop(tag, None)
    if delete is None:
        return False

    #write out new dict
    KeyValList = []
    for k,v in s3Tags.items():
        kvs={}
        kvs['Key']   = k
        kvs['Value'] = v
        KeyValList.append(kvs)

    TagSet = { 'TagSet': KeyValList }

    put_tags_response = s3_client.put_object_tagging(
        Bucket=s3bucket,
        Key=s3object,
        Tagging=TagSet
    )

    print(str(put_tags_response))

    #for key in put_tags_response['ResponseMetadata']:

    status_code = put_tags_response['ResponseMetadata']['HTTPStatusCode']

    if int(status_code) == 200:
        return True
    else:
        return False

    #return True
    #if delete:
    #    return True
    #else:
    #    return False
    #return True



def extract_rekognition_words(rekognition_json_content):

    data = json.loads(rekognition_json_content)

    List = []

    for key in data['Labels']:
        List.append(str(key['Name']))

#order matters...
# dict
#{'Name': 'Tree', 'Confidence': 90.44393920898438, 'Instances': [], 'Parents': [{'Name': 'Plant'}]}
#{'Name': 'Plant', 'Confidence': 90.44393920898438, 'Instances': [], 'Parents': []}
#{'Name': 'Urban', 'Confidence': 85.79068756103516, 'Instances': [], 'Parents': []}
#{'Name': 'Building', 'Confidence': 74.22905731201172, 'Instances': [], 'Parents': []}
#{'Name': 'Flagstone', 'Confidence': 72.1240005493164, 'Instances': [], 'Parents': []}
#{'Name': 'Vegetation', 'Confidence': 71.17005920410156, 'Instances': [], 'Parents': [{'Name': 'Plant'}]}
#{'Name': 'Garage', 'Confidence': 66.02606964111328, 'Instances': [], 'Parents': []}
#{'Name': 'Housing', 'Confidence': 65.3967056274414, 'Instances': [], 'Parents': [{'Name': 'Building'}]}
#{'Name': 'Tarmac', 'Confidence': 64.69490814208984, 'Instances': [], 'Parents': []}
#{'Name': 'Asphalt', 'Confidence': 64.69490814208984, 'Instances': [], 'Parents': []}
#{'Name': 'Home Decor', 'Confidence': 62.33171081542969, 'Instances': [], 'Parents': []}
#{'Name': 'Walkway', 'Confidence': 58.68953323364258, 'Instances': [], 'Parents': [{'Name': 'Path'}]}
#{'Name': 'Path', 'Confidence': 58.68953323364258, 'Instances': [], 'Parents': []}
#{'Name': 'Suburb', 'Confidence': 57.164859771728516, 'Instances': [], 'Parents': [{'Name': 'Urban'}, {'Name': 'Building'}]}
#{'Name': 'Bush', 'Confidence': 56.60989761352539, 'Instances': [], 'Parents': [{'Name': 'Vegetation'}, {'Name': 'Plant'}]}
#{'Name': 'Slate', 'Confidence': 56.16239547729492, 'Instances': [], 'Parents': []}
#{'Name': 'Concrete', 'Confidence': 55.33432388305664, 'Instances': [], 'Parents': []}

#[
#  "Tree", 
#  "Plant", 
#  "Urban", 
#  "Building", 
#  "Flagstone", 
#  "Vegetation", 
#  "Garage", 
#  "Housing", 
#  "Tarmac", 
#  "Asphalt", 
#  "Home Decor", 
#  "Walkway", 
#  "Path", 
#  "Suburb", 
#  "Bush", 
#  "Slate", 
#  "Concrete"
#]

    return List


def server():
    app.run(port=8880, debug=False)    


if __name__ == "__main__":
    server()


#https://docs.aws.amazon.com/code-samples/latest/catalog/python-s3-s3_basics-demo_bucket_basics.py.html
#https://docs.aws.amazon.com/rekognition/latest/dg/labels-detect-labels-image.html

#https://boto3.amazonaws.com/v1/documentation/api/latest/guide/credentials.html


#https://boto3.amazonaws.com/v1/documentation/api/latest/guide/error-handling.html


